RemcoSmitsDev/zed
# Make breakpoints persistent across Zed sessions

Release Notes:

This PR aims to make breakpoint persistence across different sessions of zed & gets zed to send breakpoints from unopened files to debug adapters when necessary.

- [x] Serialize breakpoint to be able to save them in Zed's database
- [x] Save breakpoints in zed's database
- [x] Send breakpoints from unopened files to debug adapters when necessary
- [x] Handle breakpoint overlaps with code action button
- [x] Handle breakpoint overlaps with code runner button
- [x] Color line numbers debugger.accent color if they have a breakpoint on them
- [x] Add debugger.accent color to themes, allowing users to change breakpoint colors
- [x] Add 'save_breakpoints' setting
- [x] Fix bug where toggle breakpoint action wouldn't remove breakpoints added through clicking on the gutter
- [x] Send breakpoints from unopened files to debug bug adapters
- [x] Figure out how we want to handle a the case of a line deletion where the line directly correspond to a breakpoint
- [x] Enable rendering and toggling of breakpoints within a multi buffer


Note: While this PR is close to being completed I still want time to refine and clean up how I've serialized breakpoints within Zed. I also want to try a different approach to serializing breakpoints that should allow more flexibility in the future

Future Work:
- Change breakpoint rendering to create an element instead of using an icon (Would enable right click on breakpoints)
- Add different types of breakpoints (log, hit, conditional, etc)
- Fix bug where breakpoints don't show in a multi buffer if there isn't a singular buffer already open that contains the breakpoints
    - This will probably involve making buffer_store responsible for handing breakpoints state from open -> close & close_open
- Fix multi buffer bug where toggling a breakpoint within a multi buffer with an except_id other than 1 then opening the buffer associated with the breakpoint as a singleton causes zed to crash. This only happens when both the multi buffer and singleton buffer are both opened at the same time.
  
Current State:
- The below video shows breakpoint staying persistent between different instances of zed
- Line numbers with breakpoints on them are colored the breakpoint accent color
- Breakpoints handle overlaps with code action/runner symbols
- Breakpoints are visible within a multi buffer



https://github.com/user-attachments/assets/5c83b79b-2069-4ad7-85ae-b9a7cc4bce8b
